pkg/applicationserver/applicationserver.go

```diff
@@ -632,21 +863,16 @@ func (as *ApplicationServer) handleUplink(ctx context.Context, ids ttnpb.EndDevi
 				client := ttnpb.NewAsNsClient(link.conn)
 				res, err := client.DownlinkQueueList(ctx, &ids, link.callOpts...)
 				if err != nil {
-					log.WithError(err).Warn("Failed to list downlink queue for recalculation; clearing the downlink queue")
-					req := &ttnpb.DownlinkQueueRequest{
-						EndDeviceIdentifiers: ids,
-					}
-					_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
-					if err != nil {
-						log.WithError(err).Warn("Failed to clear the downlink queue; any queued items in the Network Server are invalid")
-						events.Publish(evtInvalidQueueDataDown(ctx, ids, err))
-					} else {
-						events.Publish(evtLostQueueDataDown(ctx, ids, err))
+					logger.WithError(err).Warn("Failed to list downlink queue for recalculation; clearing the downlink queue")
+					as.resetInvalidDownlinkQueue(ctx, ids, link)
+				} else {
+					previousQueue, _ := ttnpb.PartitionDownlinksBySessionKeyIDEquality(previousSession.SessionKeyID, res.Downlinks...)
+					if err := as.recalculateDownlinkQueue(ctx, dev, link, previousSession, previousQueue, 1); err != nil {
+						logger.WithError(err).Warn("Failed to recalculate downlink queue; items lost")
```

Comment:
Should there be a lost event here?

Reply:
`recalculateDownlinkQueue` calls `runDownlinkQueueTransaction` which will call `resetInvalidDownlinkQueue` on failure thus issuing the events.